h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java

```diff
@@ -38,37 +39,82 @@ public String str() {
-    else throw new IllegalArgumentException("Expected numbers/strings. Got: " + asts[2].getClass());
-
-    final double nomatch = asts[3].exec(env).getNum();
-
-    final String[] strsTable = strsTable2;
-    final double[] dblsTable = dblsTable2;
-
-    Frame rez = new MRTask() {
-      @Override
-      public void map(Chunk c, NewChunk n) {
-        String[] domain = c.vec().domain();
-        double x;
-        int rows = c._len;
-        for (int r = 0; r < rows; ++r) {
-          x = c.isNA(r) ? nomatch : (strsTable == null ? in(dblsTable, c.atd(r), nomatch) : in(strsTable, domain[(int) c.at8(r)], nomatch));
-          n.addNum(x);
-        }
+
+    if ((fr.numCols() != 1) || ! (fr.anyVec().isCategorical() || fr.anyVec().isString()))
+      throw new IllegalArgumentException("can only match on a single categorical/string column.");
+
+    final MRTask<?> matchTask;
+    double noMatch = asts[3].exec(env).getNum();
+
+    if (asts[2] instanceof AstNumList) {
+      matchTask = new NumMatchTask(((AstNumList) asts[2]).sort().expand(), noMatch);
+    }  else if (asts[2] instanceof AstNum) {
+      matchTask = new NumMatchTask(new double[]{asts[2].exec(env).getNum()}, noMatch);
+    } else if (asts[2] instanceof AstStrList) {
+      String[] values = ((AstStrList) asts[2])._strs;
+      Arrays.sort(values);
+      matchTask = fr.anyVec().isString() ? new StrMatchTask(values, noMatch) : new CatMatchTask(values, noMatch);
+    } else if (asts[2] instanceof AstStr) {
+      String[] values = new String[]{asts[2].exec(env).getStr()};
+      matchTask = fr.anyVec().isString() ? new StrMatchTask(values, noMatch) : new CatMatchTask(values, noMatch);
+    } else
+      throw new IllegalArgumentException("Expected numbers/strings. Got: " + asts[2].getClass());
+
+    Frame result = matchTask.doAll(Vec.T_NUM, fr.anyVec()).outputFrame();
+    return new ValFrame(result);
+  }
+
+  private static class StrMatchTask extends MRTask<CatMatchTask> {
+    String[] _values;
+    double _noMatch;
+    StrMatchTask(String[] values, double noMatch) {
+      _values = values;
+      _noMatch = noMatch;
+    }
+    @Override
+    public void map(Chunk c, NewChunk nc) {
+      BufferedString bs = new BufferedString();
+      int rows = c._len;
+      for (int r = 0; r < rows; r++) {
+        double x = c.isNA(r) ? _noMatch : in(_values, c.atStr(bs, r).toString(), _noMatch);
+        nc.addNum(x);
+      }
+    }
+  }
+
+  private static class CatMatchTask extends MRTask<CatMatchTask> {
+    String[] _values;
+    double _noMatch;
+    CatMatchTask(String[] values, double noMatch) {
+      _values = values;
+      _noMatch = noMatch;
+    }
+    @Override
+    public void map(Chunk c, NewChunk nc) {
+      String[] domain = c.vec().domain();
+      int rows = c._len;
+      for (int r = 0; r < rows; r++) {
+        double x = c.isNA(r) ? _noMatch : in(_values, domain[(int) c.at8(r)], _noMatch);
```

Comment:
You can cache the results of `in(_values, domain[(int) c.at8(r)], _noMatch)`, which gets evaluated only as many times as there are different domain values. In fact, it may be better to compute the results outside of the matchtask and then simply pass a set of cat levels for which there is a match.